File: sdk/python/examples/bip32_keypair.py
#!/usr/bin/env python

#
# Shows how to use Bip32 interface to derive keys.
#
# note: this example is *not* generating keys comaptible with wallet
#

from symbolchain.Bip32 import Bip32
from symbolchain.facade.SymbolFacade import SymbolFacade


def derive_key(root_node, facade, change, index):
	path = [44, facade.BIP32_COIN_ID, 0, change, index]

	child_node = root_node.derive_path(path)
	child_key_pair = facade.bip32_node_to_key_pair(child_node)

	print(f' PATH: {path}')
	print(f' * private key: {child_key_pair.private_key}')
	print(f' *  public key: {child_key_pair.public_key}')
	address = facade.network.public_key_to_address(child_key_pair.public_key)
	print(f' *     address: {address}')
	print()


def main():
	facade = SymbolFacade('testnet')

	bip = Bip32(facade.BIP32_CURVE_NAME)
	root_node = bip.from_mnemonic('axis buzz cycle dynamic eyebrow future gym hybrid ivory just know lyrics', 'correcthorsebatterystaple')

	for change in [0, 1]:
		for index in range(3):
			derive_key(root_node, facade, change, index)


if __name__ == '__main__':
	main()

File: sdk/javascript/examples/bip32_keypair.js
#!/usr/bin/env node

//
// Shows how to use Bip32 interface to derive keys.
//
// note: this example is *not* generating keys compatible with wallet
//

import { Bip32 } from '../src/index.js';
import { SymbolFacade } from '../src/symbol/index.js';

(() => {
	const deriveKey = (rootNode, facade, change, index) => {
		const path = facade.bip32Path(0);
		path[path.length - 2] = change;
		path[path.length - 1] = index;

		const childNode = rootNode.derivePath(path);
		const childKeyPair = facade.constructor.bip32NodeToKeyPair(childNode);

		console.log(` PATH: [${path.join(', ')}]`);
		console.log(` * private key: ${childKeyPair.privateKey}`);
		console.log(` *  public key: ${childKeyPair.publicKey}`);

		const address = facade.network.publicKeyToAddress(childKeyPair.publicKey);
		console.log(` *     address: ${address}`);
		console.log('');
	};

	const facade = new SymbolFacade('testnet');

	const bip = new Bip32(facade.static.BIP32_CURVE_NAME);
	const rootNode = bip.fromMnemonic(
		'cat swing flag economy stadium alone churn speed unique patch report train',
		'correcthorsebatterystaple'
	);

	[0, 1].forEach(change => {
		[0, 1, 2].forEach(index => {
			deriveKey(rootNode, facade, change, index);
		});
	});
})();
